cmd/kubeadm/app/apis/kubeadm/v1alpha1/conversion_test.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1_test
+
+import (
+	"testing"
+
+	"k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	"k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha1"
+)
+
+func TestUpgradeBootstrapTokens(t *testing.T) {
+	testcases := []struct {
+		name        string
+		in          *v1alpha1.MasterConfiguration
+		out         *kubeadm.MasterConfiguration
+		expectedOut *kubeadm.MasterConfiguration
+		expectError bool
+	}{
+		{
+			name: "empty configs should create at least one token",
+			in:   &v1alpha1.MasterConfiguration{},
+			out:  &kubeadm.MasterConfiguration{},
+			expectedOut: &kubeadm.MasterConfiguration{
+				BootstrapTokens: []kubeadm.BootstrapToken{
+					kubeadm.BootstrapToken{
+						Token: nil,
+					},
+				},
+			},
+			expectError: false,
+		},
+		{
+			name: "fail at parsing incoming token",
+			in: &v1alpha1.MasterConfiguration{
+				Token: "some fake token",
+			},
+			expectError: true,
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			err := v1alpha1.UpgradeBootstrapTokens(tc.in, tc.out)
+
+			if tc.expectError {
+				if err == nil {
+					t.Fatal("expected an error but did not get one.")
+				}
+				// do not continue if we got an expected error
+				return
+			}
+
+			if len(tc.out.BootstrapTokens) != len(tc.expectedOut.BootstrapTokens) {
```

Comment:
Use `reflect.DeepEqual` instead.

Reply:
I think these should match exactly in testing. By doing that we don't have to create `n` assertion `if` blocks.

Reply:
I think I'm only like a 3/10 on caring about this. I suspect you are higher than me, so I'll use deep equals.